### PR TITLE
Fix Issue 20964

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

**Comment by UplinkCoder** [comment]
@9il for completeness sake you should also test a case in which the index would be out of bounds.
to see what happens when that happens.
I assume that nothing good will happen but you might be lucky and it's handled gracefully in the cases you are testing.

**Comment by nordlow** [comment]
> _No description provided._

I'm very happy to see your first dmd PR, @9il!

File: src/dmd/ctfeexpr.d [hunk 1]
```diff
@@ -873,7 +873,9 @@ UnionExp pointerArithmetic(const ref Loc loc, TOK op, Type type, Expression eptr
         error(loc, "CTFE internal error: bad pointer operation");
         goto Lcant;
     }
-    if (indx < 0 || len < indx)
+    // We allow indx to be equal to the len: one position below the right bound.
+    // So we do for the left bound for backward iteration algorithms.
+    if (indx < -1 || indx > sinteger_t(len))
     {
         error(loc, "cannot assign pointer to index %lld inside memory block `[0..%lld]`", indx, len);
```
Comment (nordlow) [inline src/dmd/ctfeexpr.d:880]: You need to add a test prefferably in `test/fail_compilation/fix20964.d` that exercises this line.

File: test/compilable/fix20964.d [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// https://issues.dlang.org/show_bug.cgi?id=20964
+
+uint foo(uint[3] m)
+{
+    auto p = m.ptr;
+    p += 3; // ok, p refers to the next right element after the block!
+    p -= 4; // ok, p refers to the next left element before the block!
```
Comment (nordlow) [inline test/compilable/fix20964.d:7]: You need to change this to

```d
p -= 1;
```

in order for this file to compile.

Comment (UplinkCoder) [reply test/compilable/fix20964.d:7]: No he doesn't.
he wants to point 1 before the pointer.
((m.ptr +3) - 4) == m.ptr - 1

**Comment by 9il** [comment]
> I'm very happy to see your first dmd PR, @9il!

Man, it isn't first.

**Comment by nordlow** [comment]
> > I'm very happy to see your first dmd PR, @9il!
> 
> Man, it isn't first.

Ahh, sorry. Only searched your open PRs.

File: test/compilable/fix20964.d (hunk 2 above)
Comment (nordlow) [reply test/compilable/fix20964.d:7]: Ahh, sorry.

**Comment by RazvanN7** [comment]
@9il this seems to be failing the testsuite

**Pull request closed without merging**